### ISSUE TITLE
batch loss fluctuates when post-training with specific domain data

### ISSUE THREAD

**Issue by Qiuzhuang** [howardhsu/BERT-for-RRC-ABSA#2]
05/14/2019 11:20:05 - INFO - root -   step 23200 batch_loss 0.665586
05/14/2019 11:21:57 - INFO - root -   step 23250 batch_loss 1.060201
05/14/2019 11:23:49 - INFO - root -   step 23300 batch_loss 1.449273
05/14/2019 11:25:40 - INFO - root -   step 23350 batch_loss 0.857883
05/14/2019 11:27:32 - INFO - root -   step 23400 batch_loss 1.117658
05/14/2019 11:29:24 - INFO - root -   step 23450 batch_loss 1.289947
05/14/2019 11:31:15 - INFO - root -   step 23500 batch_loss 1.296999
05/14/2019 11:33:07 - INFO - root -   step 23550 batch_loss 1.051140
05/14/2019 11:34:59 - INFO - root -   step 23600 batch_loss 1.439625
05/14/2019 11:36:50 - INFO - root -   step 23650 batch_loss 0.884657
05/14/2019 11:38:42 - INFO - root -   step 23700 batch_loss 1.163339
05/14/2019 11:40:33 - INFO - root -   step 23750 batch_loss 1.378097
05/14/2019 11:42:24 - INFO - root -   step 23800 batch_loss 1.190969
05/14/2019 11:44:16 - INFO - root -   step 23850 batch_loss 1.233999
05/14/2019 11:46:07 - INFO - root -   step 23900 batch_loss 1.124069
05/14/2019 11:47:59 - INFO - root -   step 23950 batch_loss 1.473860
05/14/2019 11:49:50 - INFO - root -   step 24000 batch_loss 1.483228
05/14/2019 11:51:41 - INFO - root -   step 24050 batch_loss 2.123017
05/14/2019 11:53:33 - INFO - root -   step 24100 batch_loss 1.340707
05/14/2019 11:55:24 - INFO - root -   step 24150 batch_loss 0.837390
05/14/2019 11:57:15 - INFO - root -   step 24200 batch_loss 1.205331
05/14/2019 11:59:07 - INFO - root -   step 24250 batch_loss 1.183027
05/14/2019 12:00:58 - INFO - root -   step 24300 batch_loss 1.083940

05/14/2019 12:02:49 - INFO - root -   step 24350 batch_loss 1.123584
05/14/2019 12:04:41 - INFO - root -   step 24400 batch_loss 0.431832
05/14/2019 12:06:33 - INFO - root -   step 24450 batch_loss 1.084355

05/14/2019 12:08:24 - INFO - root -   step 24500 batch_loss 0.644578
05/14/2019 12:10:16 - INFO - root -   step 24550 batch_loss 0.603636

05/14/2019 12:12:08 - INFO - root -   step 24600 batch_loss 1.205426
05/14/2019 12:13:59 - INFO - root -   step 24650 batch_loss 1.036945
05/14/2019 12:15:51 - INFO - root -   step 24700 batch_loss 1.233405
05/14/2019 12:17:43 - INFO - root -   step 24750 batch_loss 1.002707

Here is the list of parameters:

train_batch_size=16
gradient_accumulation_steps=2
num_train_steps=40000
domain data set ~ 1 million sentences

Any clues about this?

Thanks.

**Comment by Qiuzhuang** [comment]
I use https://github.com/huggingface/pytorch-pretrained-BERT/tree/master/examples/lm_finetuning
and it seems that loss is decreasing with steps.  Are there any significant difference between yours with huggingface's?

**Comment by howardhsu** [comment]
Thanks for pointing out. I'm quite busy in the coming week but I'll examine that code and see if it has better performance so to update ours.

The batch_loss is just on 16 examples so I expect to have a certain degree of fluctuation at a later stage of training but the overall loss for a long period is still decreasing.

That code is not available when we make ours at first. So we derive ours from fine-tuning code mostly to have consistent hyperparameters.

**Comment by Qiuzhuang** [comment]
Thanks very much.  For now I use huggingface's solution and looks good. 

BTW, reading your paper help us our current project a lot.

**Comment by howardhsu** [comment]
I think the variance is caused by the smaller training batch size. A larger evaluation set is expected for evaluating convergence but I don't wish to waste the small laptop corpus :-(. Also, larger accumulation step is better.
